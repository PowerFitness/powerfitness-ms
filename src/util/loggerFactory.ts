import bunyan from 'bunyan';
import { Request } from 'express';
import { ParsedQs } from 'qs';

interface ParamsDictionary {
    [key: string]: string;
}
export type LogObject = Record<string, string | number | boolean | null | ParamsDictionary | ParsedQs | Record<string, never>>;

export class Logger {
	name: string;
	req: Request | null;
	log: bunyan;
	constructor(name: string, req?: Request) {
		this.name = name;
		this.req = req || null;
		this.log = bunyan.createLogger({ name });
	}
	static _getErrorLogObject(error: Error): LogObject {
		const { message, stack } = error;
		let src: string = '';
		if (stack) {
			const splitStack: Array<string> = stack.split('\n');
			if (splitStack.length > 1) {
				src = `${splitStack[0]} ${splitStack[1]}`
			}
		}
		return { message, src };
	}

	_addExtraRequestInfo(logObject: LogObject): LogObject {
		return {
			url: this.req?.originalUrl || null,
			method: this.req?.method || null,
			params: this.req?.params || null,
			query: this.req?.query || null,
			...logObject
		}
	}

	_logByMethod(method: 'info' | 'debug' | 'error', errorOrLogObject: LogObject | Error, logObject?: LogObject): boolean | void {
		if (errorOrLogObject.constructor === Error) {
			const errorLogObject: LogObject = Logger._getErrorLogObject(errorOrLogObject as Error);
			if (logObject) {
				return this.log[method](this._addExtraRequestInfo({ ...logObject, ...errorLogObject }))
			}
			return this.log[method](this._addExtraRequestInfo(errorLogObject))
		} else {
			return this.log[method](this._addExtraRequestInfo(errorOrLogObject as LogObject));
		}
	}

	info(logOjbect: LogObject): void;
	info(error: Error): void;
	info(error: Error, logObject?: LogObject): void;

	// Implementation
	info(errorOrLogObject: LogObject | Error, logObject?: LogObject): void {
		this._logByMethod('info', errorOrLogObject, logObject)
	}

	// Overloads
	debug(logObject: LogObject): void;
	debug(error: Error): void;
	debug(error: Error, logObject?: LogObject): void;

	// Implementations
	debug(errorOrLogObject: LogObject | Error, logObject?: LogObject): void {
		this._logByMethod('debug', errorOrLogObject, logObject)
	}

	// Overloads
	error(logOjbect: LogObject): void;
	error(error: Error): void;
	error(error: Error, logObject?: LogObject): void;

	// Implementation
	error(errorOrLogObject: LogObject | Error, logObject?: LogObject): void {
		this._logByMethod('error', errorOrLogObject, logObject)
	}
}

export default (name: string, req?: Request): Logger => new Logger(name, req);
