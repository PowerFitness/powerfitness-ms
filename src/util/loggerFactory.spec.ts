import { Request } from 'express';
import loggerFactory, { Logger, LogObject } from './loggerFactory';
import Bunyan from 'bunyan';

const mockInfo: jest.Mock<void> = jest.fn();
const mockError: jest.Mock<void> = jest.fn();
const mockDebug: jest.Mock<void> = jest.fn();
const mockCreateLogger = (): Bunyan => ({
	info: (input: Record<string, never>) => mockInfo(input),
	error: (input: Record<string, never>) => mockError(input),
	debug: (input: Record<string, never>) => mockDebug(input)
}) as unknown as Bunyan;
jest.mock('bunyan', () => ({
	createLogger: () => mockCreateLogger(),
	esModule: true
}))

describe('Logger', () => {
	describe('_addExtraRequestInfo', () => {
		it('adds request details to the logObject', () => {
			const logObject: LogObject = {
				a: 'a'
			}
			const req: Request = {
				originalUrl: '/test',
				method: 'GET',
				params: { p1: 'p1' },
				query: { q1: 'q1' }
			} as unknown as Request;
			const logger: Logger = new Logger('test', req);

			const output: LogObject = logger._addExtraRequestInfo(logObject);

			expect(output).toEqual({
				url: '/test',
				method: 'GET',
				params: { p1: 'p1' },
				query: { q1: 'q1' },
				a: 'a'
			})
		})
	})
	describe('_getErrorLogObject', () => {
		it('dissects the Error object so that it can be logged', () => {
			const error: Error = { message: 'oops!', stack: 'Error: oops!\n    at test.ts:1:1\n' } as unknown as Error;
			expect(Logger._getErrorLogObject(error)).toMatchSnapshot();
		})
		it('can handle when the stack is missing', () => {
			const error: Error = { message: 'oops!' } as unknown as Error;
			expect(Logger._getErrorLogObject(error)).toMatchSnapshot();
		})
	})
	describe('_logByMethod', () => {
		// beforeEach(() => {
		// 	mockInfo.mockReset();
		// 	mockDebug.mockReset();
		// 	mockError.mockReset();
		// })
		it('logs errors', () => {
			const error: Error = {
				message: 'oops!',
				stack: 'Error: oops!\n    at test.ts:1:1\n',
				constructor: Error
			} as unknown as Error;
			const req: Request = {
				originalUrl: '/test',
				method: 'GET',
				params: { p1: 'p1' },
				query: { q1: 'q1' }
			} as unknown as Request;
			const logger: Logger = new Logger('test', req);

			logger._logByMethod('info', error);

			expect(mockInfo).toHaveBeenCalledWith({
				message: 'oops!',
				method: 'GET',
				params: {
					p1: 'p1'
				},
				query: { q1: 'q1' },
				src: 'Error: oops!     at test.ts:1:1',
				url: '/test'
			});
		})
		it('logs errors with additional logObject', () => {
			const error: Error = {
				message: 'oops!',
				stack: 'Error: oops!\n    at test.ts:1:1\n',
				constructor: Error
			} as unknown as Error;
			const logObject: LogObject = { a: 'a' };
			const req: Request = {
				originalUrl: '/test',
				method: 'GET',
				params: { p1: 'p1' },
				query: { q1: 'q1' }
			} as unknown as Request;
			const logger: Logger = new Logger('test', req);

			logger._logByMethod('info', error, logObject);

			expect(mockInfo).toHaveBeenCalledWith({
				message: 'oops!',
				method: 'GET',
				params: {
					p1: 'p1'
				},
				query: { q1: 'q1' },
				src: 'Error: oops!     at test.ts:1:1',
				url: '/test'
			});
		})
		it('logs error without stack', () => {
			const error: Error = {
				message: 'oops!',
				constructor: Error
			} as unknown as Error;
			const req: Request = {
				originalUrl: '/test',
				method: 'GET',
				params: { p1: 'p1' },
				query: { q1: 'q1' }
			} as unknown as Request;
			const logger: Logger = new Logger('test', req);

			logger._logByMethod('info', error);

			expect(mockInfo).toHaveBeenCalledWith({
				message: 'oops!',
				method: 'GET',
				params: {
					p1: 'p1'
				},
				query: { q1: 'q1' },
				src: '',
				url: '/test'
			});
		})
		it('logs error with empty stack', () => {
			const error: Error = {
				message: 'oops!',
				stack: '',
				constructor: Error
			} as unknown as Error;
			const req: Request = {
				originalUrl: '/test',
				method: 'GET',
				params: { p1: 'p1' },
				query: { q1: 'q1' }
			} as unknown as Request;
			const logger: Logger = new Logger('test', req);

			logger._logByMethod('info', error);

			expect(mockInfo).toHaveBeenCalledWith({
				message: 'oops!',
				method: 'GET',
				params: {
					p1: 'p1'
				},
				query: { q1: 'q1' },
				src: '',
				url: '/test'
			});
		})
		it('logs logObject', () => {
			const logObject: LogObject = { a: 'a' }
			const req: Request = {
				originalUrl: '/test',
				method: 'GET',
				params: { p1: 'p1' },
				query: { q1: 'q1' }
			} as unknown as Request;
			const logger: Logger = new Logger('test', req);

			logger._logByMethod('info', logObject);

			expect(mockInfo).toHaveBeenCalledWith({
				method: 'GET',
				params: {
					p1: 'p1'
				},
				query: { q1: 'q1' },
				a: 'a',
				url: '/test'
			});
		})
	})
	describe('info', () => {
		it('logs via info', () => {
			const logObject: LogObject = { a: 'a' }
			const req: Request = {
				originalUrl: '/test',
				method: 'GET',
				params: { p1: 'p1' },
				query: { q1: 'q1' }
			} as unknown as Request;
			const logger: Logger = new Logger('test', req);

			logger.info(logObject);

			expect(mockInfo).toHaveBeenCalledWith({
				method: 'GET',
				params: {
					p1: 'p1'
				},
				query: { q1: 'q1' },
				a: 'a',
				url: '/test'
			});
		})
	})
	describe('debug', () => {
		it('logs via debug', () => {
			const logObject: LogObject = { a: 'a' }
			const req: Request = {
				originalUrl: '/test',
				method: 'GET',
				params: { p1: 'p1' },
				query: { q1: 'q1' }
			} as unknown as Request;
			const logger: Logger = new Logger('test', req);

			logger.debug(logObject);

			expect(mockDebug).toHaveBeenCalledWith({
				method: 'GET',
				params: {
					p1: 'p1'
				},
				query: { q1: 'q1' },
				a: 'a',
				url: '/test'
			});
		})
	})
	describe('error', () => {
		it('logs via error', () => {
			const logObject: LogObject = { a: 'a' }
			const req: Request = {
				originalUrl: '/test',
				method: 'GET',
				params: { p1: 'p1' },
				query: { q1: 'q1' }
			} as unknown as Request;
			const logger: Logger = new Logger('test', req);

			logger.error(logObject);

			expect(mockError).toHaveBeenCalledWith({
				method: 'GET',
				params: {
					p1: 'p1'
				},
				query: { q1: 'q1' },
				a: 'a',
				url: '/test'
			});
		})
	})
})
describe('loggerFactory', () => {
	it('creates a new Logger', () => {
		const logObject: LogObject = { a: 'a' }
		const req: Request = {
			originalUrl: '/test',
			method: 'GET',
			params: { p1: 'p1' },
			query: { q1: 'q1' }
		} as unknown as Request;
		const logger: Logger = loggerFactory('test', req);

		logger.info(logObject);

		expect(mockInfo).toHaveBeenCalledWith({
			method: 'GET',
			params: {
				p1: 'p1'
			},
			query: { q1: 'q1' },
			a: 'a',
			url: '/test'
		});
	})
	// it('does not need a request object', () => {
	// 	const logObject: LogObject = { a: 'a' }
	// 	const logger: Logger = loggerFactory('test');

	// 	logger.info(logObject);

	// 	expect(mockInfo).toHaveBeenCalledWith({ a: 'a' });
	// })
})
