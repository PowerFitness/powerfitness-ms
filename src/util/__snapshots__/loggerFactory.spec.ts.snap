// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logger _getErrorLogObject can handle when the stack is missing 1`] = `
Object {
  "message": "oops!",
  "src": "",
}
`;

exports[`Logger _getErrorLogObject dissects the Error object so that it can be logged 1`] = `
Object {
  "message": "oops!",
  "src": "Error: oops!     at test.ts:1:1",
}
`;
